### PR TITLE
Drag-and-drop interaction bar widgets

### PR DESCRIPTION
**Pull request by Sjmarf**
Adds the ability to rearrange the interaction bar widgets. There are separate layouts for the post and comment interaction bars. The Info Stack must be part of the user's layout, and there cannot be too many widgets in the stack.

The widget layouts are stored in a JSON file.

### CODE REVIEW

**Comment by JakeShirley** [comment]
FYI building this locally has a few different Swift errors on my machine.

**Comment by Sjmarf** [comment]
> FYI building this locally has a few different Swift errors on my machine.

Hmmm. This is what someone said last time. I'm on the latest beta... maybe that's causing it

**Comment by JakeShirley** [comment]
This is the main error I am seeing.  Extra sus that CI doesn't seem to be running on your PR either

image

**Comment by EricBAndrews** [comment]
I'm seeing those exact same errors

XCode Version 14.3.1 (14E300c)
MacOS Ventura 13.1 (22C65)

**Comment by EricBAndrews** [comment]
First and foremost: fantastic job, this is an amazing feature that is going to absolutely thrill our users.

It looks like Swift doesn't like those let declarations of view building functions. I was able to get it to run with some extremely crude and slapdash refactoring into functions, though ye be warned that this is pretty ugly code: https://github.com/EricBAndrews/mlemapp/commit/4be88591af522aeb1cd83b93092602a67565603b

Once it compiles and I can test it thoroughly I'll give it a proper review. I will say as a pre-review that `LayoutWidgetEditView.swift` is going to need a whole lot of those pyramids broken down and moved into functions anyway--there's a lot going on in there already, so if the view structure itself isn't *squeaky* clean that file is going to become a nightmare to maintain

**Comment by JakeShirley** [comment]
Some small design feedback: We might want to render a preview of what the "gap" section shows (stats like post time, etc.).

**Comment by Sjmarf** [comment]
> Some small design feedback: We might want to render a preview of what the "gap" section shows (stats like post time, etc.).
> 


Yeah, I considered this. However, the interaction bar as it appears in the editor is much larger than in the real thing - there is extra padding around each widget, which allows you to pick them up easier. This means that the Info Stack is much shorter than how it actually appears in the feed, and so the info stack content probably wont fit properly if I try to render it there.

**Comment by Sjmarf** [comment]
> First and foremost: fantastic job, this is an amazing feature that is going to absolutely thrill our users.
> 
> 
> 
> It looks like Swift doesn't like those let declarations of view building functions. I was able to get it to run with some extremely crude and slapdash refactoring into functions, though ye be warned that this is pretty ugly code: https://github.com/EricBAndrews/mlemapp/commit/4be88591af522aeb1cd83b93092602a67565603b
> 
> 
> 
> Once it compiles and I can test it thoroughly I'll give it a proper review. I will say as a pre-review that `LayoutWidgetEditView.swift` is going to need a whole lot of those pyramids broken down and moved into functions anyway--there's a lot going on in there already, so if the view structure itself isn't *squeaky* clean that file is going to become a nightmare to maintain

Thanks, I'll refactor it 👍

**Comment by Sjmarf** [comment]
It should build now (hopefully).

**Comment by Sjmarf** [comment]
Ok, I've done some refactoring. It should build now, hopefully.

File: Mlem/Views/Tabs/Settings/Components/Views/Appearance/Shared/LayoutWidgetEditView.swift [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                    }
+            )
+        
+            if widgetModel.shouldShowDraggingWidget {
+                HStack {
+                    LayoutWidgetView(widget: widgetModel.widgetDragging!, animation: animation)
+                }
+                .frame(
+                    width: widgetModel.widgetDragging!.rect!.width,
+                    height: widgetModel.widgetDragging!.rect!.height
```
Comment (EricBAndrews) [inline Mlem/Views/Tabs/Settings/Components/Views/Appearance/Shared/LayoutWidgetEditView.swift:110]: *nit*

force unwraps

**Review by EricBAndrews — CHANGES_REQUESTED** [review]
Lots of nits, mostly focused on avoiding force unwrapping, but overall looks good!

File: Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        for widget in items where widget.rect != nil {
+            if widget.rect?.contains(location) ?? false {
+                return widget
+            }
+        }
+        return nil
+    }
+    
+    func isValidDropLocation(_ widgetDragging: LayoutWidget? = nil) -> Bool {
+        costLimit == .infinity || totalCost(widgetDragging) <= costLimit
```
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift:39]: It's for optimisation, yeah. I'm assuming `||` works in the same way as in Python? I'll test it when I get time later.

Ie: if the first part of the `||` expression evaluates to `false`, the result cannot possibly be `true`, and so there's no need to evaluate the second part. This saves calling the `totalCost` function

File: Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    func totalCost(_ widgetDragging: LayoutWidget? = nil) -> Float {
+        self.items.reduce(0) { accumulator, element in
+            accumulator + element.type.cost
+        } + (widgetDragging?.type.cost ?? 0)
+    }
+    
+    func getItemDictionary() -> [LayoutWidgetType: LayoutWidget] {
+        var widgets = [LayoutWidgetType: LayoutWidget]()
+        for widget in itemsToRender {
+            widgets[widget!.type] = widget
```
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift:51]: Sure 👍

File: Mlem/Views/Tabs/Settings/Components/Views/Appearance/Post/PostSettingsView.swift [hunk 4]
```diff
@@ -39,28 +38,38 @@ struct PostSettingsView: View {
...
+                Button {
+                    showingWidgetSheet = true
+                } label: {
+                    HStack {
+                        Text("Customize Widgets")
+                        Spacer()
+                        Image(systemName: "arrow.up.square.fill")
+                        Image(systemName: "arrow.down.square")
+                    }
+                    .foregroundStyle(.pink)
```
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Views/Tabs/Settings/Components/Views/Appearance/Post/PostSettingsView.swift:62]: I originally had it as a nav page, but changed it to a sheet for some reason that I now can't remember xD

I'll try changing it back and let you know if I encounter an issue

File: Mlem/Views/Tabs/Settings/Components/Views/Appearance/Shared/LayoutWidgetEditView.swift [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        
+        for widget in widgets {
+            barWidgets.append(.init(widget))
+        }
+        
+        var trayWidgets: [LayoutWidget] = .init()
+        
+        for widget in LayoutWidgetType.allCases where !widgets.contains(widget) {
+            trayWidgets.append(.init(widget))
+        }
```
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Views/Tabs/Settings/Components/Views/Appearance/Shared/LayoutWidgetEditView.swift:56]: 😎 

This is my first Swift project so I'm slowly learning all this fancy stuff as I go haha.

File: Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    func drop(_ widget: LayoutWidget) {
+        self.items.append(widget)
+    }
+    
+    func getItemAtLocation(_ location: CGPoint) -> LayoutWidget? {
+        for widget in items where widget.rect != nil {
+            if widget.rect?.contains(location) ?? false {
+                return widget
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift:34]: Close 😉 

`items.first { $0.rect?.contains(location) ?? false }`

File: Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Models/Settings/LayoutWidgets/LayoutWidgetCollection.swift:39]: Yep, that is indeed how it works

File: Mlem/Views/Tabs/Settings/Components/Views/Appearance/Post/PostSettingsView.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Views/Tabs/Settings/Components/Views/Appearance/Post/PostSettingsView.swift]: I got it working 👍

**Pull request merged**